plugins/modules/proxmox_vm_info.py

```diff
@@ -154,42 +168,46 @@ def get_vms_from_cluster_resources(self):
                 msg="Failed to retrieve VMs information from cluster resources: %s" % e
             )
 
-    def get_vms_from_nodes(self, vms_unfiltered, type, vmid=None, name=None, node=None):
-        vms = []
-        for vm in vms_unfiltered:
-            if (
-                type != vm["type"]
-                or (node and vm["node"] != node)
-                or (vmid and int(vm["vmid"]) != vmid)
-                or (name is not None and vm["name"] != name)
-            ):
-                continue
-            vms.append(vm)
-        nodes = frozenset([vm["node"] for vm in vms])
-        for node in nodes:
-            if type == "qemu":
-                vms_from_nodes = self.proxmox_api.nodes(node).qemu().get()
-            else:
-                vms_from_nodes = self.proxmox_api.nodes(node).lxc().get()
-            for vmn in vms_from_nodes:
-                for vm in vms:
-                    if int(vm["vmid"]) == int(vmn["vmid"]):
-                        vm.update(vmn)
-                        vm["vmid"] = int(vm["vmid"])
-                        vm["template"] = proxmox_to_ansible_bool(vm["template"])
-                        break
-
-        return vms
-
-    def get_qemu_vms(self, vms_unfiltered, vmid=None, name=None, node=None):
+    def get_vms_from_nodes(self, cluster_machines, type, vmid=None, name=None, node=None, config=None):
+        # Leave in dict only machines that user wants to know about
+        filtered_vms = {
+            vm: info for vm, info in cluster_machines.items() if not (
+                type != info["type"]
+                or (node and info["node"] != node)
+                or (vmid and int(info["vmid"]) != vmid)
+                or (name is not None and info["name"] != name)
+            )
+        }
+        # Get list of unique node names and loop through it to get info about machines.
+        nodes = frozenset([info["node"] for vm, info in filtered_vms.items()])
+        for this_node in nodes:
+            # "type" is mandatory and can have only values of "qemu" or "lxc". Seems that use of reflection is safe.
+            call_vm_getter = getattr(self.proxmox_api.nodes(this_node), type)
+            vms_from_this_node = call_vm_getter.get()
```

Comment:
The original code ran
```suggestion
            vms_from_this_node = call_vm_getter().get()
```
instead.

Reply:
Thank you for help :) I'm fighting against debugging unit testing of module for whole two days. Will try your suggestions.